### ISSUE TITLE
client always reconnect if have many connect number 

### ISSUE THREAD

**Issue by jyzhangbo** [emqx/emqx#2561]
#### Environment

- OS: CentOS Linux release 7.3.1611 (Core)
- Erlang/OTP: 
- EMQ: emqx-centos7-v3.1.0.zip

#### Description
when the connect number is a few, it work well,but if the number more than dozens,the client always reconnect. But when change the version to emqttd-centos7-v2.3.11, it work well.

The 3.1.0 conf:
```
cluster.name=emqxcl-gateway
Value:Enum
cluster.discovery=static
cluster.autoheal=on
cluster.autoclean=5m
cluster.static.seeds=gateway1@172.21.1.9,gateway2@172.21.2.10,gateway3@172.21.2.9
node.name=gateway1@172.21.1.9
node.cookie=emqxsecretcookie
node.data_dir=data
node.async_threads=32
node.process_limit=2048000
node.max_ports=1024000
node.dist_buffer_size=8MB
node.max_ets_tables=256000
node.fullsweep_after=1000
node.crash_dump=log/crash.dump
node.proto_dist=inet_tcp
node.dist_net_ticktime=60
node.dist_listen_min=6369
node.dist_listen_max=6369
rpc.tcp_server_port=5369
rpc.tcp_client_port=5369
rpc.connect_timeout=5s
rpc.send_timeout=5s
rpc.authentication_timeout=5s
rpc.call_receive_timeout=15s
rpc.socket_keepalive_idle=900s
rpc.socket_keepalive_interval=75s
rpc.socket_keepalive_count=9
log.to=both
log.level=error
log.dir=log
log.file=emqx.log
log.chars_limit=1024
log.rotation.size=10MB
log.rotation.count=5
#log.info.file=info.log
#log.error.file=error.log
allow_anonymous=true
acl_nomatch=allow
acl_file=etc/acl.conf
enable_acl_cache=on
acl_cache_max_size=32
acl_cache_ttl=1m
acl_deny_action=ignore
mqtt.max_packet_size=1MB
mqtt.max_clientid_len=65535
mqtt.max_topic_levels=0
mqtt.max_qos_allowed=2
mqtt.max_topic_alias=0
mqtt.retain_available=true
mqtt.wildcard_subscription=true
mqtt.shared_subscription=true
mqtt.ignore_loop_deliver=false
zone.external.idle_timeout=15s
zone.external.enable_acl=on
zone.external.enable_ban=on
zone.external.enable_stats=on
zone.external.acl_deny_action=ignore
zone.external.force_gc_policy=1000|1MB
zone.external.keepalive_backoff=0.75
zone.external.max_subscriptions=0
zone.external.upgrade_qos=off
zone.external.max_inflight=32
zone.external.retry_interval=20s
zone.external.max_awaiting_rel=100
zone.external.await_rel_timeout=300s
zone.external.session_expiry_interval=2h
zone.external.max_mqueue_len=1000
zone.external.mqueue_priorities=none
zone.external.mqueue_default_priority=highest
zone.external.mqueue_store_qos0=true
zone.external.enable_flapping_detect=off
zone.external.flapping_threshold=10,1m
zone.external.flapping_banned_expiry_interval=1h
zone.external.use_username_as_clientid=false
zone.internal.allow_anonymous=true
zone.internal.enable_stats=on
zone.internal.enable_acl=off
zone.internal.acl_deny_action=ignore
zone.internal.max_subscriptions=0
zone.internal.max_inflight=32
zone.internal.max_awaiting_rel=100
zone.internal.max_mqueue_len=1000
zone.internal.mqueue_store_qos0=true
zone.internal.enable_flapping_detect=off
zone.internal.flapping_threshold=10,1m
zone.internal.flapping_banned_expiry_interval=1h
zone.internal.use_username_as_clientid=false
listener.ssl.external=7883
listener.ssl.external.acceptors=64
listener.ssl.external.max_connections=1024000
listener.ssl.external.max_conn_rate=5000
listener.ssl.external.active_n=100
listener.ssl.external.zone=external
listener.ssl.external.access.1=allowall
listener.ssl.external.handshake_timeout=15s
listener.ssl.external.keyfile=etc/certs/key.pem
listener.ssl.external.certfile=etc/certs/cert.pem
listener.ssl.external.ciphers=ECDHE-ECDSA-AES256-GCM-SHA384,ECDHE-RSA-AES256-GCM-SHA384,ECDHE-ECDSA-AES256-SHA384,ECDHE-RSA-AES256-SHA384,ECDHE-ECDSA-DES-CBC3-SHA,ECDH-ECDSA-AES256-GCM-SHA384,ECDH-RSA-AES256-GCM-SHA384,ECDH-ECDSA-AES256-SHA384,ECDH-RSA-AES256-SHA384,DHE-DSS-AES256-GCM-SHA384,DHE-DSS-AES256-SHA256,AES256-GCM-SHA384,AES256-SHA256,ECDHE-ECDSA-AES128-GCM-SHA256,ECDHE-RSA-AES128-GCM-SHA256,ECDHE-ECDSA-AES128-SHA256,ECDHE-RSA-AES128-SHA256,ECDH-ECDSA-AES128-GCM-SHA256,ECDH-RSA-AES128-GCM-SHA256,ECDH-ECDSA-AES128-SHA256,ECDH-RSA-AES128-SHA256,DHE-DSS-AES128-GCM-SHA256,DHE-DSS-AES128-SHA256,AES128-GCM-SHA256,AES128-SHA256,ECDHE-ECDSA-AES256-SHA,ECDHE-RSA-AES256-SHA,DHE-DSS-AES256-SHA,ECDH-ECDSA-AES256-SHA,ECDH-RSA-AES256-SHA,AES256-SHA,ECDHE-ECDSA-AES128-SHA,ECDHE-RSA-AES128-SHA,DHE-DSS-AES128-SHA,ECDH-ECDSA-AES128-SHA,ECDH-RSA-AES128-SHA,AES128-SHA
#listener.ssl.external.psk_ciphers=PSK-AES128-CBC-SHA,PSK-AES256-CBC-SHA,PSK-3DES-EDE-CBC-SHA,PSK-RC4-SHA
listener.ssl.external.reuseaddr=true
#bridge.azure.psk_ciphers=PSK-AES128-CBC-SHA,PSK-AES256-CBC-SHA,PSK-3DES-EDE-CBC-SHA,PSK-RC4-SHA
module.presence=on
module.presence.qos=1
module.subscription=off
module.rewrite=off
plugins.etc_dir=etc/plugins/
plugins.loaded_file=data/loaded_plugins
plugins.expand_plugins_dir=plugins/
broker.sys_interval=1m
broker.enable_session_registry=on
broker.session_locking_strategy=quorum
broker.shared_subscription_strategy=random
broker.shared_dispatch_ack_enabled=false
broker.route_batch_clean=off
sysmon.long_gc=false
sysmon.long_schedule=240
sysmon.large_heap=8MB
sysmon.busy_port=false
sysmon.busy_dist_port=true
os_mon.cpu_check_interval=60s
os_mon.cpu_high_watermark=80%
os_mon.cpu_low_watermark=60%
os_mon.mem_check_interval=60s
os_mon.sysmem_high_watermark=70%
os_mon.procmem_high_watermark=5%
vm_mon.check_interval=30s
vm_mon.process_high_watermark=80%
vm_mon.process_low_watermark=60%
```

The error log:
```
2019-05-21 04:18:01.455 [debug] D4EE076B7806@114.240.145.184:45670 [RuleEngine] Client(D4EE076B7806) connected, connack: 0
2019-05-21 04:18:01.456 [debug] D4EE076B7806@114.240.145.184:45670 [Protocol] SEND CONNACK(Q0, R0, D0, AckFlags=0, ReasonCode=0)
2019-05-21 04:18:03.336 [debug] D4EE076B7806@114.240.145.184:45670 [Connection] RECV <<48,9,0,6,111,110,108,105,110,101,49>>
2019-05-21 04:18:03.336 [debug] D4EE076B7806@114.240.145.184:45670 [Protocol] RECV PUBLISH(Q0, R0, D0, Topic=online, PacketId=undefined, Payload=<<"1">>)
2019-05-21 04:18:03.337 [info] D4EE076B7806@114.240.145.184:45670 PUBLISH to online: <<"1">>
2019-05-21 04:18:03.337 [debug] D4EE076B7806@114.240.145.184:45670 [RuleEngine] Publish Message(Id=
  peername => {{114,240,145,184},45670},
  username => <<"root">>})
2019-05-21 04:18:03.337 [debug] D4EE076B7806@114.240.145.184:45670 [RuleEngine] Message dropped for no subscription: Message(Id=ow_publish => true,
  peername => {{114,240,145,184},45670},
  username => <<"root">>})
2019-05-21 04:18:03.385 [debug] D4EE076B7806@114.240.145.184:45670 [Connection] RECV <<48,9,0,6,111,110,108,105,110,101,49>>
2019-05-21 04:18:03.385 [debug] D4EE076B7806@114.240.145.184:45670 [Protocol] RECV PUBLISH(Q0, R0, D0, Topic=online, PacketId=undefined, Payload=<<"1">>)
2019-05-21 04:18:03.385 [info] D4EE076B7806@114.240.145.184:45670 PUBLISH to online: <<"1">>
2019-05-21 04:18:03.386 [debug] D4EE076B7806@114.240.145.184:45670 [RuleEngine] Publish Message(Id=
  peername => {{114,240,145,184},45670},
  username => <<"root">>})
2019-05-21 04:18:03.386 [debug] D4EE076B7806@114.240.145.184:45670 [RuleEngine] Message dropped for no subscription: Message(Id=w_publish => true,
  peername => {{114,240,145,184},45670},
  username => <<"root">>})
2019-05-21 04:18:05.486 [debug] D4EE076B7806@114.240.145.184:45670 [Connection] RECV <<224,0>>
2019-05-21 04:18:05.486 [debug] D4EE076B7806@114.240.145.184:45670 [Protocol] RECV DISCONNECT(Q0, R0, D0, ReasonCode=0)
2019-05-21 04:18:05.487 [debug] D4EE076B7806@114.240.145.184:45670 [Connection] Terminated for normal
2019-05-21 04:18:05.487 [info] D4EE076B7806@114.240.145.184:45670 [Protocol] Shutdown for normal
2019-05-21 04:18:05.487 [debug] D4EE076B7806@114.240.145.184:45670 [RuleEngine] Client(D4EE076B7806) disconnected, reason_code: normal
2019-05-21 04:18:07.763 [debug] D4EE076B7806@114.240.145.184:45673 [RuleEngine] Client(D4EE076B7806) connected, connack: 0
2019-05-21 04:18:07.763 [debug] D4EE076B7806@114.240.145.184:45673 [Protocol] SEND CONNACK(Q0, R0, D0, AckFlags=0, ReasonCode=0)
2019-05-21 04:18:07.775 [debug] D4EE076B7806@114.240.145.184:45673 [Connection] RECV <<130,64,0,1,0,59,47,99,111,109,47,122,105,114,111,111,109,
                    47,105,111,116,47,122,103,97,116,101,119,97,121,47,100,48,
                    102,57,54,52,54,48,100,49,48,57,52,100,102,56,98,48,101,56,
                    51,49,52,55,101,48,54,57,50,49,57,50,47,35,1>>
2019-05-21 04:18:07.775 [debug] D4EE076B7806@114.240.145.184:45673 [Protocol] RECV SUBSCRIBE(Q1, R0, D0, PacketId=1, TopicFilters=[{<<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#">>,
                           #{nl => 0,qos => 1,rap => 0,rc => 0,rh => 0}}])
2019-05-21 04:18:07.775 [debug] D4EE076B7806@114.240.145.184:45673 [RuleEngine] Client(D4EE076B7806) will subscribe: [{<<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#">>,
                                                    #{nl => 0,qos => 1,
                                                      rap => 0,rc => 1,
                                                      rh => 0}}]
2019-05-21 04:18:10.338 [debug] D4EE076B7806@114.240.145.184:45673 [Connection] RECV <<50,99,0,31,47,99,111,109,47,122,105,114,111,111,109,47,105,
                    111,116,47,122,103,97,116,101,119,97,121,47,111,110,108,
                    105,110,101,0,2,123,10,9,34,116,111,107,101,110,34,58,9,34,
                    100,48,102,57,54,52,54,48,100,49,48,57,52,100,102,56,98,48,
                    101,56,51,49,52,55,101,48,54,57,50,49,57,50,34,44,10,9,34,
                    115,116,97,116,117,101,34,58,9,34,49,34,10,125>>
2019-05-21 04:18:10.338 [debug] D4EE076B7806@114.240.145.184:45673 [Protocol] RECV PUBLISH(Q1, R0, D0, Topic=/com/ziroom/iot/zgateway/online, PacketId=2, Payload=<<"{\n\t\"token\":\t\"d0f96460d1094df8b0e83147e
0692192\",\n\t\"statue\":\t\"1\"\n}">>)
2019-05-21 04:18:10.338 [info] D4EE076B7806@114.240.145.184:45673 PUBLISH to /com/ziroom/iot/zgateway/online: <<"{\n\t\"token\":\t\"d0f96460d1094df8b0e83147e0692192\",\n\t\"statue\":\t\"1\"\n}">>
2019-05-21 04:18:10.339 [debug] D4EE076B7806@114.240.145.184:45673 [RuleEngine] Publish Message(Id=ish => true,
  peername => {{114,240,145,184},45673},
  username => <<"root">>})
2019-05-21 04:18:10.339 [debug] D4EE076B7806@114.240.145.184:45673 [Protocol] SEND PUBACK(Q0, R0, D0, PacketId=2)
2019-05-21 04:18:10.439 [debug] D4EE076B7806@114.240.145.184:45673 [Connection] RECV <<48,9,0,6,111,110,108,105,110,101,49>>
2019-05-21 04:18:10.439 [debug] D4EE076B7806@114.240.145.184:45673 [Protocol] RECV PUBLISH(Q0, R0, D0, Topic=online, PacketId=undefined, Payload=<<"1">>)
2019-05-21 04:18:10.439 [info] D4EE076B7806@114.240.145.184:45673 PUBLISH to online: <<"1">>
2019-05-21 04:18:10.439 [debug] D4EE076B7806@114.240.145.184:45673 [RuleEngine] Publish Message(Id=
  peername => {{114,240,145,184},45673},
  username => <<"root">>})
2019-05-21 04:18:10.439 [debug] D4EE076B7806@114.240.145.184:45673 [RuleEngine] Message dropped for no subscription: Message(Id=_publish => true,
  peername => {{114,240,145,184},45673},
  username => <<"root">>})
2019-05-21 04:18:10.488 [debug] D4EE076B7806@114.240.145.184:45673 [Connection] RECV <<48,9,0,6,111,110,108,105,110,101,49>>
2019-05-21 04:18:10.488 [debug] D4EE076B7806@114.240.145.184:45673 [Protocol] RECV PUBLISH(Q0, R0, D0, Topic=online, PacketId=undefined, Payload=<<"1">>)
2019-05-21 04:18:10.489 [info] D4EE076B7806@114.240.145.184:45673 PUBLISH to online: <<"1">>
2019-05-21 04:18:10.489 [debug] D4EE076B7806@114.240.145.184:45673 [RuleEngine] Publish Message(Id=
  peername => {{114,240,145,184},45673},
  username => <<"root">>})
2019-05-21 04:18:10.489 [debug] D4EE076B7806@114.240.145.184:45673 [RuleEngine] Message dropped for no subscription: Message(Id=w_publish => true,
  peername => {{114,240,145,184},45673},
  username => <<"root">>})
2019-05-21 04:18:12.777 [info] D4EE076B7806 PUBLISH to /com/ziroom/iot/zgateway/online: <<"{\n\t\"token\":\t\"d0f96460d1094df8b0e83147e0692192\",\n\t\"statue\":\t\"0\"\n}">>
2019-05-21 04:18:12.777 [debug] D4EE076B7806 [RuleEngine] Publish Message(Id= => <<"root">>})
2019-05-21 04:18:12.777 [error] D4EE076B7806 ** Generic server <0.16907.52> terminating 
** Last message in was {'$gen_cast',
                           {subscribe,<0.16655.52>,
                               {1,undefined,
                                [{<<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#">>,
                                  #{nl => 0,qos => 1,rap => 0,rc => 1,
                                    rh => 0}}]}}}
** When Server state == {state,15000,true,#Fun<emqx_session.1.98926271>,
                            <<"D4EE076B7806">>,<<"root">>,<0.16655.52>,
                            undefined,1,0,#{},false,
                            {emqx_inflight,32,{0,nil}},
                            20000,undefined,
                            {mqueue,true,1000,0,0,none,infinity,
                                {queue,[],[],0}},
                            #{},100,300000,undefined,0,undefined,true,
                            undefined,
                            {emqx_gc,
                                #{cnt => {1000,1000},
                                  oct => {1048576,1048576}}},
                            {1558,383487,758826},
                            {message,
                                <<0,5,137,87,110,54,33,52,169,74,52,0,65,15,0,0>>,
                                1,<<"D4EE076B7806">>,
                                #{dup => false,retain => false},
                                #{username => <<"root">>},
                                <<"/com/ziroom/iot/zgateway/online">>,
                                <<"{\n\t\"token\":\t\"d0f96460d1094df8b0e83147e0692192\",\n\t\"statue\":\t\"0\"\n}">>,
                                {1558,383487,754570}},
                            undefined}
** Reason for termination == 
** {{timeout,{gen_server,call,
                         [<0.2541.0>,
                          {subscribe,<<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#">>}]}},
    [{gen_server,call,2,[{file,"gen_server.erl"},{line,215}]},
     {emqx_session,do_subscribe,5,[{file,"src/emqx_session.erl"},{line,1152}]},
     {emqx_session,'-handle_cast/2-fun-0-',4,
                   [{file,"src/emqx_session.erl"},{line,491}]},
     {emqx_session,handle_cast,2,[{file,"src/emqx_session.erl"},{line,488}]},
     {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,637}]},
     {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,711}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
2019-05-21 04:18:12.782 [error] D4EE076B7806 crasher:
    initial call: emqx_session:init/1
    pid: <0.16907.52>
    registered_name: []
    exception exit: {timeout,
                        {gen_server,call,
                            [<0.2541.0>,
                             {subscribe,
                                 <<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#">>}]}}
      in function  gen_server:call/2 (gen_server.erl, line 215)
      in call from emqx_session:do_subscribe/5 (src/emqx_session.erl, line 1152)
      in call from emqx_session:'-handle_cast/2-fun-0-'/4 (src/emqx_session.erl, line 491)
      in call from emqx_session:handle_cast/2 (src/emqx_session.erl, line 488)
      in call from gen_server:try_dispatch/4 (gen_server.erl, line 637)
      in call from gen_server:handle_msg/6 (gen_server.erl, line 711)
    ancestors: [emqx_session_sup,emqx_sm_sup,emqx_sup,<0.2479.0>]
    message_queue_len: 1
    messages: [{'EXIT',<0.16655.52>,
                      {timeout,
                          {gen_server,call,
                              [<0.2541.0>,
                               {subscribe,
                                   <<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#">>}]}}}]
    links: [<0.2551.0>]
    dictionary: [{'$logger_metadata$',#{client_id => <<"D4EE076B7806">>}},
                  {force_shutdown_policy,
                      #{max_heap_size => 0,message_queue_len => 0}}]
    trap_exit: true
    status: running
    heap_size: 6772
    stack_size: 27
    reductions: 204353
  neighbours:
2019-05-21 04:18:13.008 [debug] D4EE076B7806@114.240.145.184:16015 [RuleEngine] Client(D4EE076B7806) connected, connack: 0
2019-05-21 04:18:13.009 [debug] D4EE076B7806@114.240.145.184:16015 [Protocol] SEND CONNACK(Q0, R0, D0, AckFlags=0, ReasonCode=0)

```

*A description of the issue*

**Comment by jyzhangbo** [comment]
Would you give a  general outline of this bug.Is it a matter of some parameters?

**Comment by HJianBo** [comment]
Hi, @jyzhangbo Can you give the host performance and the exact amount of connected client?

**Comment by jyzhangbo** [comment]
@HJianBo Hi，the host performance is  8C16G.The connected client about 200 on one computer.These 200 are connected at the same time，then 90% start reconnecting.

The clusters conf id:
```
cluster.discovery=static
cluster.autoheal=on
cluster.autoclean=5m
cluster.static.seeds=gateway1@172.21.1.9,gateway2@172.21.2.10,gateway3@172.21.2.9
```

**Comment by tradingtrace** [comment]
Hi @jyzhangbo 

I reproduced the issue you mentioned above. The test environment is a 2-core 4G cloud server. Two nodes run in docker, one runs from source compiled.

I guess you launched all about 600 connections in a very short time. And after connected, a subscription request was sent immediately. 

For `Emqx`, it is very easy to process 600 connections. But to process 600 global subscriptions, it has to take some time. 

In order to make it easier to match topics at the time of publication, `Emqx` implemented a `Trie` tree in Mnesia with a table named `emqx_trie`, which holds all subscriptions. A database transaction will be used when the `emqx_trie` table is updated, which protects consistency but slows down the processing.

According your logs, each client will subscribe to a specific topic when it is connected:

```erlang
Topic = <<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#">>.
```

The topic consists of the same prefix and the hash of the client. It has a total of 7 levels. 

```erlang
(emqx@172.17.0.1)7> length(emqx_topic:triples(Topic)).
7
```

To subscribe to such a topic, **7 records** of `emqx_trie` table must be updated. And because of the same prefix, 5 of 7 records are the same. That is, all subscriptions want to lock the same 5 records, which creates competition and slows down the `emqx_broker` process.

I printed the `emqx_broker` process_info during the test. Most of the time the `emqx_broker` process is waiting for a lock or waiting to retry the transaction:

```erlang
emqx_broker_1: [{current_function,{timer,sleep,1}},{message_queue_len,49}]
emqx_broker_2: [{current_function,{timer,sleep,1}},{message_queue_len,38}]
emqx_broker_3: [{current_function,{mnesia_locker,receive_wlocks,4}},{message_queue_len,59}]
emqx_broker_4: [{current_function,{mnesia_locker,receive_wlocks,4}},{message_queue_len,47}]
```

I think based on the current version `Emqx`, we can do:

1. Make connections and subscriptions on a time interval. After connecting, randomly wait for a certain period of time before starting a subscription. For example, In my test (300 connections per node), let the client randomly sleep for 0~10 seconds after connected, and then request the subscription. This could avoid the above issue.
2. Minimize the level of the topic if possible. For example, changes `/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#` to  `com.ziroom.iot.zgateway/d0f96460d1094df8b0e83147e0692192/#`, which changes topic level from 7 to 3, and the number of competitive records changes from 5 to 1, it will also improve the above issue.

PS:

**In v3.0 `Emqx` all subscriptions are global (a database transaction occurs on all nodes)**, which means a client can receive the publications of a topic from all nodes but only needs to subscribe to one node.

**But in v2.0 `Emqx` all subscriptions are local by default (only updates the local node ets)**, which means a client can receive only the publications from the local node. 


@HJianBo 

Is it necessary to update `emqx_route` and `emqx_trie` in the same transactions? If not, what about this:

1. Change `emqx_trie` from mnesia table to ets table.
2. Subscribe the simple `mnesia_table_event` of `emqx_route`
3. Update `emqx_trie` ets when the `mnesia_table_event` is received (only when wildcard topic is written or deleted).

**Comment by tradingtrace** [comment]
Addition:

3. Use wildcards with caution. The topic will update `emqx_trie` with a transaction only if it contains wildcard. So if we have a limited number of topics, for example:
    ```erlang
    <<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/a">>
    <<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/b">>
    <<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/c">>
    ```
    It's better to subscribe them explicitly insted of subscribing this:
     ```erlang
    <<"/com/ziroom/iot/zgateway/d0f96460d1094df8b0e83147e0692192/#">>
    ```

**Comment by jyzhangbo** [comment]
@tradingtrace Thank you very much for your reply,

**Comment by u007** [comment]
i faced the same with websocket, it keep reconnecting every 2 to 4s

**Comment by zmstone** [comment]
> i faced the same with websocket, it keep reconnecting every 2 to 4s

hi @u007 if you find such issue in newer version. Feel free to create a new issue. Would be great if you can share the logs when creating.

**Comment by u007** [comment]
sorry its same

**Comment by zmstone** [comment]
> sorry its same

Please create a new issue by following the issue template.
